File: native/bindings/endstone/inventory/inventory.cpp
#include "inventory.h"

#include "item_stack.h"

#include <endstone/inventory/inventory.h>
#include <endstone/inventory/item_stack.h>

#include <optional>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace aegilex::native::inventory {

namespace {

[[nodiscard]] bool valid_inventory_type_id(const std::string_view type_id) noexcept
{
    if (type_id.empty()) {
        return false;
    }
    std::size_t colons = 0;
    for (const auto byte : type_id) {
        const auto value = static_cast<std::uint8_t>(byte);
        const auto valid = (value >= static_cast<std::uint8_t>('a') && value <= static_cast<std::uint8_t>('z')) ||
                           (value >= static_cast<std::uint8_t>('0') && value <= static_cast<std::uint8_t>('9')) ||
                           value == static_cast<std::uint8_t>('_') || value == static_cast<std::uint8_t>(':');
        if (!valid) {
            return false;
        }
        colons += value == static_cast<std::uint8_t>(':') ? 1U : 0U;
    }
    return colons <= 1 && type_id.front() != ':' && type_id.back() != ':';
}

} // namespace

class Inventory::impl {
  public:
    explicit impl(endstone::Inventory *inventory) noexcept : inventory(inventory)
    {
    }

    endstone::Inventory *inventory;
};

Inventory::Inventory(endstone::Inventory *inventory) noexcept : impl(std::make_shared<class Inventory::impl>(inventory))
{
}

endstone::Inventory *Inventory::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->inventory;
}

std::int32_t Inventory::getSize() const
{
    try {
        return impl->inventory->getSize();
    }
    catch (...) {
        return 0;
    }
}

std::int32_t Inventory::getMaxStackSize() const
{
    try {
        return impl->inventory->getMaxStackSize();
    }
    catch (...) {
        return 0;
    }
}

std::unique_ptr<ItemStack> Inventory::getItem(const std::int32_t index) const
{
    try {
        if (index < 0) {
            return std::unique_ptr<ItemStack>();
        }
        const auto item = impl->inventory->getItem(index);
        return item.has_value() ? std::unique_ptr<ItemStack>(new ItemStack(*item)) : std::unique_ptr<ItemStack>();
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void Inventory::setItem(const std::int32_t index, const ItemStack &item) const
{
    try {
        if (index < 0 || item.native() == nullptr) {
            return;
        }
        impl->inventory->setItem(index, std::optional<endstone::ItemStack>{*item.native()});
    }
    catch (...) {
    }
}

void Inventory::clear() const
{
    try {
        impl->inventory->clear();
    }
    catch (...) {
    }
}

void Inventory::clearIndex(const std::int32_t index) const
{
    try {
        if (index < 0) {
            return;
        }
        impl->inventory->clear(index);
    }
    catch (...) {
    }
}

std::unique_ptr<ItemStack> Inventory::addItem(const ItemStack &item) const
{
    try {
        if (item.native() == nullptr) {
            return std::unique_ptr<ItemStack>();
        }
        auto leftovers = impl->inventory->addItem(std::vector<endstone::ItemStack>{*item.native()});
        const auto leftover = leftovers.find(0);
        return leftover != leftovers.end() ? std::unique_ptr<ItemStack>(new ItemStack(leftover->second))
                                           : std::unique_ptr<ItemStack>();
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

std::unique_ptr<ItemStack> Inventory::removeItem(const ItemStack &item) const
{
    try {
        if (item.native() == nullptr) {
            return std::unique_ptr<ItemStack>();
        }
        auto leftovers = impl->inventory->removeItem(std::vector<endstone::ItemStack>{*item.native()});
        const auto leftover = leftovers.find(0);
        return leftover != leftovers.end() ? std::unique_ptr<ItemStack>(new ItemStack(leftover->second))
                                           : std::unique_ptr<ItemStack>();
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

bool Inventory::containsType(const rust::Str type_id) const
{
    try {
        const std::string_view identifier(type_id.data(), type_id.size());
        return valid_inventory_type_id(identifier) && impl->inventory->contains(std::string(identifier));
    }
    catch (...) {
        return false;
    }
}

bool Inventory::containsStack(const ItemStack &item) const
{
    try {
        return item.native() != nullptr && impl->inventory->contains(*item.native());
    }
    catch (...) {
        return false;
    }
}

bool Inventory::containsAtLeastType(const rust::Str type_id, const std::int32_t amount) const
{
    try {
        const std::string_view identifier(type_id.data(), type_id.size());
        return valid_inventory_type_id(identifier) && impl->inventory->containsAtLeast(std::string(identifier), amount);
    }
    catch (...) {
        return false;
    }
}

bool Inventory::containsAtLeastStack(const ItemStack &item, const std::int32_t amount) const
{
    try {
        return item.native() != nullptr && impl->inventory->containsAtLeast(*item.native(), amount);
    }
    catch (...) {
        return false;
    }
}

std::int32_t Inventory::firstType(const rust::Str type_id) const
{
    try {
        const std::string_view identifier(type_id.data(), type_id.size());
        return valid_inventory_type_id(identifier) ? impl->inventory->first(std::string(identifier)) : -1;
    }
    catch (...) {
        return -1;
    }
}

std::int32_t Inventory::firstStack(const ItemStack &item) const
{
    try {
        return item.native() != nullptr ? impl->inventory->first(*item.native()) : -1;
    }
    catch (...) {
        return -1;
    }
}

std::int32_t Inventory::firstEmpty() const
{
    try {
        return impl->inventory->firstEmpty();
    }
    catch (...) {
        return -1;
    }
}

bool Inventory::isEmpty() const
{
    try {
        return impl->inventory->isEmpty();
    }
    catch (...) {
        return false;
    }
}

void Inventory::removeType(const rust::Str type_id) const
{
    try {
        const std::string_view identifier(type_id.data(), type_id.size());
        if (valid_inventory_type_id(identifier)) {
            impl->inventory->remove(std::string(identifier));
        }
    }
    catch (...) {
    }
}

void Inventory::removeStack(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->remove(*item.native());
        }
    }
    catch (...) {
    }
}

} // namespace aegilex::native::inventory

File: rust/crates/aegilex-runtime/src/host/endstone/inventory/inventory.rs
//! Core ABI implementation for `native/bindings/endstone/inventory/inventory.h`.

use crate::host::endstone::inventory::resources::*;
use crate::host::endstone::support::*;

impl crate::core_host::imports::HostInventory for PluginStoreState {
    fn inventory_get_size(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.get-size")?;
            self.handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.getSize())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)
        })())
    }

    fn inventory_get_item(
        &mut self,
        self_: u32,
        index: i32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.get-item")?;
            let invocation_id = self.invocation_id;
            let item = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?
                .getItem(index);
            if item.is_null() {
                return Ok(None);
            }
            Ok(Some(
                self.insert_item_stack_resource(item)
                    .map_err(map_core_host_error)?,
            ))
        })())
    }

    fn inventory_set_item(
        &mut self,
        self_: u32,
        index: i32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.set-item")?;
            let invocation_id = self.invocation_id;
            let inventory = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            if let Some(item) = item {
                let handle = item_stack_resource_handle(self, item).map_err(map_core_host_error)?;
                let item = self
                    .handles
                    .item_stack(invocation_id, handle)
                    .ok_or_else(not_found)
                    .map_err(map_core_host_error)?;
                inventory.setItem(index, item);
            } else {
                inventory.clearIndex(index);
            }
            Ok(())
        })())
    }

    fn inventory_get_contents(
        &mut self,
        self_: u32,
    ) -> Result<Result<Vec<InventoryInventoryItem>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.get-contents")?;
            let invocation_id = self.invocation_id;
            let items = {
                let inventory = self
                    .handles
                    .inventory(
                        invocation_id,
                        inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                    )
                    .ok_or_else(not_found)
                    .map_err(map_core_host_error)?;
                (0..inventory.getSize())
                    .map(|index| (index, inventory.getItem(index)))
                    .collect::<Vec<_>>()
            };
            let mut contents = Vec::new();
            for (index, item) in items {
                let item = if item.is_null() {
                    None
                } else {
                    Some(
                        self.insert_item_stack_resource(item)
                            .map_err(map_core_host_error)?,
                    )
                };
                contents.push(InventoryInventoryItem { index, item });
            }
            Ok(contents)
        })())
    }

    fn inventory_set_contents(
        &mut self,
        self_: u32,
        items: Vec<Option<u32>>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.set-contents")?;
            let invocation_id = self.invocation_id;
            let inventory = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            if items.len() > inventory.getSize() as usize {
                return Err(map_core_host_error(not_found()));
            }
            inventory.clear();
            for (index, item) in items.into_iter().enumerate() {
                if let Some(item) = item {
                    let handle =
                        item_stack_resource_handle(self, item).map_err(map_core_host_error)?;
                    let item = self
                        .handles
                        .item_stack(invocation_id, handle)
                        .ok_or_else(not_found)
                        .map_err(map_core_host_error)?;
                    inventory.setItem(index as i32, item);
                }
            }
            Ok(())
        })())
    }

    fn inventory_clear(&mut self, self_: u32) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.clear")?;
            self.handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.clear())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn inventory_clear_index(
        &mut self,
        self_: u32,
        index: i32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.clear-index")?;
            self.handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.clearIndex(index))
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn inventory_get_max_stack_size(
        &mut self,
        self_: u32,
    ) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.get-max-stack-size")?;
            self.handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.getMaxStackSize())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)
        })())
    }

    fn inventory_add_item(
        &mut self,
        self_: u32,
        items: Vec<u32>,
    ) -> Result<Result<Vec<InventoryItemStackResult>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.add-item")?;
            let invocation_id = self.invocation_id;
            let native_leftovers = {
                let inventory = self
                    .handles
                    .inventory(
                        invocation_id,
                        inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                    )
                    .ok_or_else(not_found)
                    .map_err(map_core_host_error)?;
                let mut native_leftovers = Vec::new();
                for (slot, resource) in items.into_iter().enumerate() {
                    let handle =
                        item_stack_resource_handle(self, resource).map_err(map_core_host_error)?;
                    let item = self
                        .handles
                        .item_stack(invocation_id, handle)
                        .ok_or_else(not_found)
                        .map_err(map_core_host_error)?;
                    let leftover = inventory.addItem(item);
                    if !leftover.is_null() {
                        native_leftovers.push((slot, leftover));
                    }
                }
                native_leftovers
            };
            let mut leftovers = Vec::new();
            for (slot, leftover) in native_leftovers {
                leftovers.push(InventoryItemStackResult {
                    slot: slot as i32,
                    item: self
                        .insert_item_stack_resource(leftover)
                        .map_err(map_core_host_error)?,
                });
            }
            Ok(leftovers)
        })())
    }

    fn inventory_remove_item(
        &mut self,
        self_: u32,
        items: Vec<u32>,
    ) -> Result<Result<Vec<InventoryItemStackResult>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.remove-item")?;
            let invocation_id = self.invocation_id;
            let native_removed = {
                let inventory = self
                    .handles
                    .inventory(
                        invocation_id,
                        inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                    )
                    .ok_or_else(not_found)
                    .map_err(map_core_host_error)?;
                let mut native_removed = Vec::new();
                for (slot, resource) in items.into_iter().enumerate() {
                    let handle =
                        item_stack_resource_handle(self, resource).map_err(map_core_host_error)?;
                    let item = self
                        .handles
                        .item_stack(invocation_id, handle)
                        .ok_or_else(not_found)
                        .map_err(map_core_host_error)?;
                    let removed = inventory.removeItem(item);
                    if !removed.is_null() {
                        native_removed.push((slot, removed));
                    }
                }
                native_removed
            };
            let mut removed = Vec::new();
            for (slot, leftover) in native_removed {
                removed.push(InventoryItemStackResult {
                    slot: slot as i32,
                    item: self
                        .insert_item_stack_resource(leftover)
                        .map_err(map_core_host_error)?,
                });
            }
            Ok(removed)
        })())
    }

    fn inventory_first_slot(
        &mut self,
        self_: u32,
        request: InventoryItemRequest,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.first-slot")?;
            let invocation_id = self.invocation_id;
            let inventory = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            let (stack, type_id) = match request.matcher {
                InventoryItemMatcher::Stack(resource) => (
                    Some(
                        self.handles
                            .item_stack(
                                invocation_id,
                                item_stack_resource_handle(self, resource)
                                    .map_err(map_core_host_error)?,
                            )
                            .ok_or_else(not_found)
                            .map_err(map_core_host_error)?,
                    ),
                    None,
                ),
                InventoryItemMatcher::TypeId(type_id) => (None, Some(type_id)),
            };
            for slot in 0..inventory.getSize() {
                let item = inventory.getItem(slot);
                let Some(item_ref) = item.as_ref() else {
                    continue;
                };
                let matches = type_id
                    .as_ref()
                    .is_some_and(|type_id| item_ref.getType() == *type_id)
                    || stack.is_some_and(|stack| item_ref.equals(stack));
                if matches {
                    return Ok(Some(slot as u32));
                }
            }
            Ok(None)
        })())
    }

    fn inventory_contains(
        &mut self,
        self_: u32,
        request: InventoryItemRequest,
    ) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.contains")?;
            let invocation_id = self.invocation_id;
            let inventory = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            match request.matcher {
                InventoryItemMatcher::TypeId(type_id) => Ok(request.amount.map_or_else(
                    || inventory.containsType(&type_id),
                    |amount| inventory.containsAtLeastType(&type_id, amount),
                )),
                InventoryItemMatcher::Stack(item) => {
                    let handle =
                        item_stack_resource_handle(self, item).map_err(map_core_host_error)?;
                    let item = self
                        .handles
                        .item_stack(invocation_id, handle)
                        .ok_or_else(not_found)
                        .map_err(map_core_host_error)?;
                    Ok(request.amount.map_or_else(
                        || inventory.containsStack(item),
                        |amount| inventory.containsAtLeastStack(item, amount),
                    ))
                }
            }
        })())
    }

    fn inventory_list_matching(
        &mut self,
        self_: u32,
        request: InventoryItemRequest,
    ) -> Result<Result<Vec<InventoryItemStackResult>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.list-matching")?;
            let invocation_id = self.invocation_id;
            let matched_items = {
                let inventory = self
                    .handles
                    .inventory(
                        invocation_id,
                        inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                    )
                    .ok_or_else(not_found)
                    .map_err(map_core_host_error)?;
                let (stack, type_id) = match request.matcher {
                    InventoryItemMatcher::Stack(resource) => (
                        Some(
                            self.handles
                                .item_stack(
                                    invocation_id,
                                    item_stack_resource_handle(self, resource)
                                        .map_err(map_core_host_error)?,
                                )
                                .ok_or_else(not_found)
                                .map_err(map_core_host_error)?,
                        ),
                        None,
                    ),
                    InventoryItemMatcher::TypeId(type_id) => (None, Some(type_id)),
                };
                let mut matched_items = Vec::new();
                for slot in 0..inventory.getSize() {
                    let item = inventory.getItem(slot);
                    let Some(item_ref) = item.as_ref() else {
                        continue;
                    };
                    let matches = type_id
                        .as_ref()
                        .is_some_and(|type_id| item_ref.getType() == *type_id)
                        || stack.is_some_and(|stack| item_ref.equals(stack));
                    if matches {
                        matched_items.push((slot, item));
                    }
                }
                matched_items
            };
            let mut result = Vec::new();
            for (slot, item) in matched_items {
                result.push(InventoryItemStackResult {
                    slot,
                    item: self
                        .insert_item_stack_resource(item)
                        .map_err(map_core_host_error)?,
                });
            }
            Ok(result)
        })())
    }

    fn inventory_first_empty(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.first-empty")?;
            let slot = self
                .handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.firstEmpty())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            Ok((slot >= 0).then_some(slot as u32))
        })())
    }

    fn inventory_is_empty(&mut self, self_: u32) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.is-empty")?;
            self.handles
                .inventory(
                    self.invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.isEmpty())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)
        })())
    }

    fn inventory_remove_matching(
        &mut self,
        self_: u32,
        request: InventoryItemRequest,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "inventory.inventory.remove-matching")?;
            let invocation_id = self.invocation_id;
            let inventory = self
                .handles
                .inventory(
                    invocation_id,
                    inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            match request.matcher {
                InventoryItemMatcher::TypeId(type_id) => inventory.removeType(&type_id),
                InventoryItemMatcher::Stack(item) => {
                    let handle =
                        item_stack_resource_handle(self, item).map_err(map_core_host_error)?;
                    let item = self
                        .handles
                        .item_stack(invocation_id, handle)
                        .ok_or_else(not_found)
                        .map_err(map_core_host_error)?;
                    inventory.removeStack(item);
                }
            }
            Ok(())
        })())
    }
}
